neo/io/nwbio.py

```diff
@@ -0,0 +1,810 @@
+        # this handles only icephys_electrode for now
+        electrodes = {}
+        devices = {}
+        for segment in block.segments:
+            for signal in chain(segment.analogsignals, segment.irregularlysampledsignals):
+                if "nwb_electrode" in signal.annotations:
+                    elec_meta = signal.annotations["nwb_electrode"].copy()
+                    if elec_meta["name"] not in electrodes:
+                        # todo: check for consistency if the name is already there
+                        if elec_meta["device"]["name"] in devices:
+                            device = devices[elec_meta["device"]["name"]]
+                        else:
+                            device = nwbfile.create_device(**elec_meta["device"])
+                            devices[elec_meta["device"]["name"]] = device
+                        elec_meta.pop("device")
+                        electrodes[elec_meta["name"]] = nwbfile.create_icephys_electrode(
+                            device=device, **elec_meta
+                        )
+        return electrodes
+
+    def _write_segment(self, nwbfile, segment, electrodes):
+        # maybe use NWB trials to store Segment metadata?
+        for i, signal in enumerate(chain(segment.analogsignals, segment.irregularlysampledsignals)):
+            assert signal.segment is segment
+            if not signal.name:
+                signal.name = "%s : analogsignal%d" % (segment.name, i)
+            self._write_signal(nwbfile, signal, electrodes)
+
+        for i, train in enumerate(segment.spiketrains):
+            assert train.segment is segment
+            if not train.name:
+                train.name = "%s : spiketrain%d" % (segment.name, i)
+            self._write_spiketrain(nwbfile, train)
+
+        for i, event in enumerate(segment.events):
+            assert event.segment is segment
+            if not event.name:
+                event.name = "%s : event%d" % (segment.name, i)
+            self._write_event(nwbfile, event)
+
+        for i, epoch in enumerate(segment.epochs):
+            if not epoch.name:
+                epoch.name = "%s : epoch%d" % (segment.name, i)
+            self._write_epoch(nwbfile, epoch)
+
+    def _write_signal(self, nwbfile, signal, electrodes):
+        hierarchy = {'block': signal.segment.block.name, 'segment': signal.segment.name}
+        if "nwb_type" in signal.annotations:
+            timeseries_class = get_class(*signal.annotations["nwb_type"])
+        else:
+            timeseries_class = TimeSeries  # default
+        additional_metadata = {name[4:]: value
+                               for name, value in signal.annotations.items()
+                               if name.startswith("nwb:")}
+        if "nwb_electrode" in signal.annotations:
+            electrode_name = signal.annotations["nwb_electrode"]["name"]
+            additional_metadata["electrode"] = electrodes[electrode_name]
```

Comment:
This would only be relevant for descendants of `PatchClampSeries`. Extracellular electrophysiology analog signals form a map to electrodes in a different way. They use an `electrodes` arg that is a `DynamicTableRegion`, pointing to a region of the `electrodes` `DynamicTable`.

Reply:
thanks for the pointer.

That reminds me, at the moment the `pynwb.icephys` module has `PatchClampSeries` but no `SharpElectrodeSeries`. Should I open a ticket for this?

Reply:
sure! That would be a ticket in the nwb-schema repo

Reply:
created https://github.com/NeurodataWithoutBorders/nwb-schema/issues/479